src/gtimelog/main.py

```diff
@@ -1483,6 +1509,25 @@ def populate_log(self):
         self.add_footer()
         self.scroll_to_end()
 
+    def _get_log_order_key(self):
+        """
+        Returns a function suitable as key parameter for the ordered function
+
+        The parameter to be sorted is deemed a list item as returned by
+        TimeCollection.grouped_entries
+        """
+        if self.log_order == 'alphanum':
+            return lambda x: x[1]
+        elif self.log_order == 'duration':
+            return lambda x: x[2]
+        elif self.log_order == 'tasks':
+            return lambda x: self.tasks.index(x[1])
```

Comment:
list.index() can raise an IndexError if there's no match.  Also, it does an O(N) linear search over the entire list.

I think it'd be better to keep a dictionary mapping tasks to ordering, and then do

    return lambda x: self.task_order.get(x[1], sys.maxsize)

to put all unknown tasks at the end.

Another thing, how does this handle ties?  E.g. what about two tasks that took exactly the same time, when sorting by duration?  How should they be sorted?

Reply:
Okay, so self.tasks is a TaskList and not a list, and `index()` is a custom helper.  Disregard most of the above.

The part about handling ties still stands.

Reply:
I think I want these sort helpers to be global functions in the timelog.py module, so they can have unit tests.  Then this function can become

```python
        return {
            'alphanum': sort_by_entry_title,
            'duration': sort_by_duration,
            'tasks': sort_by_task_order(self.tasks),
            'asis': sort_by_start_time,
        }[self.log_order]
```

There's a slight asymmetry here where sort_by_task_order() is a higher order function, but I don't think it can be helped:

```python
def sort_by_entry_title(item):
    start, entry, duration = item
    return (entry, start, duration)

def sort_by_duration(item):
    start, entry, duration = item
    return (duration, start, entry)

def sort_by_start_time(item):
    return item

def sort_by_task_order(tasks):

    def sort_key(item):
        start, entry, duration = item
        return (tasks.order(entry), item)

    return sort_key
```

I've also made it clearer what these x[1], x[2] mean and added tie breaking (by falling back to the previous sort rules, which are start time, then entry title, then duration).

I should probably remember this is the 21st century and refactor TimeCollection.grouped_entries() to return named tuples.

Reply:
I would rename your functions in accordance with the new options name start-time/name/duration/task-list (or the other way around if you prefer).

Also, it's still a bit fuzzy, but I'm asking myself if `_get_log_order_key` shouldn't become a method of "window/TimeCollection", so that we simply call `work, slack = window.grouped_entries(sorted_by=self.log_order, sorted_tasks=self.tasks)` and let `grouped_entries` completely decide what it does with the information. This is slightly more complex, _but_ more encapsulated also, and it would make it easier to keep the same interface later on for `categorized_work_entries`.

Reply:
Sure, that works for me.

Reply:
Result is slightly different, but I think the spirit is the same. I've tried to capture my thoughts in the comments of the code in timelog. Next step: add tests...